enterprise/server/remote_execution/nbd/nbdclient/nbdclient.go

```diff
@@ -27,8 +30,184 @@ const (
 
 	// How often to poll for an NBD to become ready.
 	readyCheckPollInterval = 500 * time.Microsecond
+
+	// EXT4_IOC_RESIZE_FS is the ioctl constant for resizing an ext4 FS.
+	// Computed from C: https://gist.github.com/bduffany/ce9b594c2166ea1a4564cba1b5ed652d
+	EXT4_IOC_RESIZE_FS = 0x40086610
 )
 
+// nbdclient ProcessHandle commands. The parent sends these to the child, and
+// the child responds with "OK".
+//
+// We don't use gRPC for this communication because we want the NBD client
+// process to be as simple as possible in order to maximize reliability and
+// prevent stuck NBD requests.
+const (
+	initCmd    = "INIT"
+	mountCmd   = "MOUNT"
+	unmountCmd = "UNMOUNT"
+)
+
+// Process represents an nbdclient process.
+//
+// nbd clients are run in dedicated processes to prevent interference from other
+// goroutines running in the same binary. For example, if all goroutines enter a
+// syscall which depends on an NBD read (such as a read() or execve()), then it
+// will lead to deadlock, because the nbd client goroutine would not be able to
+// run.
+type Process struct {
+	name   string
+	input  *os.File
+	output *bufio.Scanner
+	err    error
+}
+
+// Start starts an nbdclient as a background process. The returned object can be
+// used to communicate with the process.
+func Start(name, mountPath string) (*Process, error) {
+	inR, inW, err := os.Pipe()
+	if err != nil {
+		return nil, status.InternalErrorf("failed to create pipe: %s", err)
+	}
+	outR, outW, err := os.Pipe()
+	if err != nil {
+		return nil, status.InternalErrorf("failed to create pipe: %s", err)
+	}
+	cmd := exec.Command(os.Args[0], append(os.Args[1:], "--nbdclient")...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.ExtraFiles = []*os.File{inR, outW}
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	// Close input reader and output writer handles; these are only used by the
+	// child.
+	inR.Close()
+	outW.Close()
+
+	scanner := bufio.NewScanner(outR)
+	h := &Process{name: name, input: inW, output: scanner}
+	go func() {
+		err := cmd.Wait()
+		h.err = status.InternalErrorf("child exited: %s", err)
```

Comment:
Do we need to check for nil err here?

Reply:
The nbdclient process should stay running forever so we want to treat any exit here as an error (added a clarifying comment)